cpp/src/arrow/status.h

```diff
@@ -363,17 +357,25 @@ class ARROW_EXPORT [[nodiscard]] Status : public util::EqualityComparable<Status
   void AddContextLine(const char* filename, int line, const char* expr);
 #endif
 
+  // Construct a Status which is not destroyed and is cheap to copy.
+  static Status MakeStatic(StatusCode code, std::string msg,
+                           std::shared_ptr<StatusDetail> detail);
  private:
   struct State {
     StatusCode code;
     std::string msg;
     std::shared_ptr<StatusDetail> detail;
+    bool is_static = false;
   };
   // OK status has a `NULL` state_.  Otherwise, `state_` points to
   // a `State` structure containing the error code and message(s)
   State* state_;
 
   void DeleteState() {
+    // ARROW-2400: On certain compilers, splitting off the slow path improves
+    // performance significantly.
+    if (ARROW_PREDICT_TRUE(state_ == NULL)) return;
+    if (ARROW_PREDICT_FALSE(state_->is_static)) return;
```

Comment:
Probably better to have a single condition here to make it obvious that the rest of the function is the cold code:

```
if (ARROW_PREDICT_TRUE(state_ == NULL || state_->is_static)) return;
```

Reply:
I agree that this is more obvious for human readers, but I think the priority here is to make it obvious to the compiler that the *hottest* condition is `state_ == NULL`, independent of any other consideration

Reply:
I'm saying it would be better to the compiler as well not only humans. What the previous code achieved with the `ARROW_PREDICT_FALSE` was telling the compiler that `DeleteState()` is code and shouldn't be inlined. Since lots of `Status` are destroyed all over the place, that reduces binary size and avoids pollution of the I-cache with cold code.

```cpp
    if (ARROW_PREDICT_FALSE(state_ != NULL)) {
      DeleteState();
    }
```

Reply:
I think I see what you mean, how about:
```suggestion
    if (ARROW_PREDICT_FALSE(state_ != NULL)) {
      if (ARROW_PREDICT_FALSE(state_->is_static)) return;
      delete state_;
```
Then it should still be clear that `State::~State()` is cold. I could also explicitly move the definition of `State::~State()` into `status.cc`, which should prevent inlining

Reply:
[aargh visibility](https://github.com/apache/arrow/actions/runs/11521170513/job/32074214669?pr=44493#step:7:998)

Reply:
I think the previous structure is the best: call `DeleteState` after a single unlikely condition. And the `is_static` check lives inside `DeleteState`.